### PR TITLE
[cuDNN][64-bit indexing] cuDNN v9.3+ supports non-batch-splittable convolutions with > 2**31 elements

### PR DESCRIPTION
**Pull request by eqy**
For longstanding issues such as #95024

cc @csarofeen @ptrblck @xwang233 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -422,7 +422,13 @@ struct ConvParams {
   // that maybe the compiledWithCuDNN() check sometimes segfaults (though I can't imagine how)
 #if !defined(C10_MOBILE)
     if (needs_64bit_indexing_no_split(input, weight)) {
-      return false;
+      static long cudnn_version = detail::getCUDAHooks().versionCuDNN();
+      if (!(cudnn_version >= 90300 && at::native::cudnnv8_enabled_check_debug())) {
+        TORCH_WARN_ONCE("cuDNN cannot be used for large non-batch-splittable convolutions"
+                        " if the V8 API is not enabled or before cuDNN version 9.3+."
```
Comment (bhack) [inline aten/src/ATen/native/Convolution.cpp:428]: Are we sure about this `or`? Or do we need `and`?

Comment (bhack) [reply aten/src/ATen/native/Convolution.cpp:428]: /cc @Skylion007

Comment (eqy) [reply aten/src/ATen/native/Convolution.cpp:428]: Are you referring to the warning message? You can apply De Morgan's law to translate the predicate to if cudnn_version < 9.3 _or_ the V8 API is disabled, then non-batch-splittable 64bit cases cannot be dispatched to cuDNN.

Comment (bhack) [reply aten/src/ATen/native/Convolution.cpp:428]: Yes but I am still entering in the condition. What is then the condition in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/ConvUtils.h#L80 ?

Comment (bhack) [reply aten/src/ATen/native/Convolution.cpp:428]: @eqy Isn't the logic wrong currently? See https://github.com/pytorch/pytorch/blob/main/c10/util/env.cpp#L93

Comment (bhack) [reply aten/src/ATen/native/Convolution.cpp:428]: Check https://github.com/pytorch/pytorch/pull/139963 if it could help

**Pull request closed without merging**